Client/game_sa/CPedSA.cpp

```diff
@@ -10,413 +10,268 @@
-                if (pWeaponModel)
+                CModelInfo* modelInfo = pGame->GetModelInfo(model);
+                if (modelInfo)
                 {
-                    pWeaponModel->Request(BLOCKING, "CPedSA::GiveWeapon");
-                    pWeaponModel->MakeCustomModel();
+                    modelInfo->Request(BLOCKING, "CPedSA::GiveWeapon");
+                    modelInfo->MakeCustomModel();
                 }
             }
+
             // If the weapon is satchels, load the detonator too
             if (weaponType == WEAPONTYPE_REMOTE_SATCHEL_CHARGE)
-            {
-                /*int iModel = pGame->GetWeaponInfo ( WEAPONTYPE_DETONATOR )->GetModel();
-                if ( iModel )
-                {
-                    CModelInfo * pWeaponModel = pGame->GetModelInfo ( iModel );
-                    if ( pWeaponModel )
-                    {
-                        pWeaponModel->Request ( true, true );
-                    }
-                }*/
-                // Load the weapon and give it properly so getPedWeapon shows the weapon is there.
                 GiveWeapon(WEAPONTYPE_DETONATOR, 1, WEAPONSKILL_STD);
-            }
         }
     }
 
-    DWORD dwReturn = 0;
-    DWORD dwFunc = FUNC_GiveWeapon;
-    DWORD dwThis = (DWORD)GetInterface();
-    _asm
-    {
-        mov     ecx, dwThis
-        push    1
-        push    uiAmmo
-        push    weaponType
-        call    dwFunc
-        mov     dwReturn, eax
-    }
-
-    CWeapon* pWeapon = GetWeapon((eWeaponSlot)dwReturn);
+    // eWeaponType __thiscall CPed::GiveWeapon(CPed *this, eWeaponType weaponID, signed int ammo, int a4)
+    // Last argument is unused
+    eWeaponSlot weaponSlot = ((eWeaponSlot(__thiscall*)(CEntitySAInterface*, std::uint8_t, std::uint32_t, std::uint8_t))FUNC_GiveWeapon)(m_pInterface, weaponType, ammo, 1);
 
-    return pWeapon;
+    CWeapon* weapon = GetWeapon(weaponSlot);
+    return weapon;
 }
 
 CWeapon* CPedSA::GetWeapon(eWeaponType weaponType)
 {
-    if (weaponType < WEAPONTYPE_LAST_WEAPONTYPE)
-    {
-        CWeapon* pWeapon = GetWeapon(pGame->GetWeaponInfo(weaponType)->GetSlot());
-        if (pWeapon->GetType() == weaponType)
-            return pWeapon;
-    }
-    return NULL;
+    if (weaponType >= WEAPONTYPE_LAST_WEAPONTYPE)
+        return nullptr;
+
+    CWeapon* weapon = GetWeapon(pGame->GetWeaponInfo(weaponType)->GetSlot());
+    if (!weapon || weapon->GetType() != weaponType)
+        return nullptr;
+
+    return weapon;
 }
 
 CWeapon* CPedSA::GetWeapon(eWeaponSlot weaponSlot)
 {
-    if (weaponSlot < WEAPONSLOT_MAX)
-    {
-        return m_pWeapons[weaponSlot];
-    }
-    else
-        return NULL;
+    return (weaponSlot >= 0 && weaponSlot < WEAPONSLOT_MAX) ? m_pWeapons[weaponSlot] : nullptr;
 }
 
 void CPedSA::ClearWeapons()
 {
     // Remove all the weapons
-    for (unsigned int i = 0; i < WEAPONSLOT_MAX; i++)
+    for (CWeaponSA* weapon : m_pWeapons)
```

Comment:
The code compiles correctly because `m_pWeapons` is a built-in array of pointers with a fixed size. Built-in arrays in C++ support range-based for loops, allowing iteration over all elements of the array, even though the array does not have classic iterators like STL containers

Reply:
Hmm, didnt know that standard arrays also have iterators